src/engine/SpatialJoinAlgorithms.cpp

```diff
@@ -374,16 +455,36 @@ Result SpatialJoinAlgorithms::LibspatialjoinAlgorithm() {
   // smaller one. Compute the bounding box of the smaller table (appropriately
   // inflated for `WITHIN_DIST` joins) and only add those geometries from the
   // larger table that intersect this bounding box.
+  auto runParser = [&](IdTableAndJoinColumn smaller,
+                       IdTableAndJoinColumn larger, bool smallerIsRight) {
+    // Parse and add all geometries of the smaller side
+    auto [boxSmall, countSmall] = libspatialjoinParse(
+        smallerIsRight, smaller, sweeper, NUM_THREADS, std::nullopt);
+
+    // Filtering by bounding box *after* parsing is only necessary if
+    // precomputed bounding boxes for filtering *before* parsing are not
+    // available.
+    if (!qec_->getIndex().getVocab().isGeoInfoAvailable()) {
+      sweeper.setFilterBox(boxSmall);
+    }
+
+    // Parse and add the relevant (intersection with the bounding box)
+    // geometries from the larger side
+    auto [boxLarge, countLarge] =
+        libspatialjoinParse(!smallerIsRight, larger, sweeper, NUM_THREADS,
+                            sweeper.getPaddedBoundingBox(boxSmall));
+
+    // If we have filtered out all geometries or one side is otherwise empty,
+    // bail out early.
+    return countSmall > 0 && countLarge > 0;
+  };
+  bool nonEmptyChildren;
   if (idTableLeft->size() < idTableRight->size()) {
-    auto box = libspatialjoinParse(false, idTableLeft, leftJoinCol, sweeper,
-                                   NUM_THREADS);
-    sweeper.setFilterBox(box);
-    libspatialjoinParse(true, idTableRight, rightJoinCol, sweeper, NUM_THREADS);
+    nonEmptyChildren = runParser({idTableLeft, leftJoinCol},
+                                 {idTableRight, rightJoinCol}, false);
```

Comment:
You can bind the `{tableLeft, leftJoinCol}`to a variable outside this `if`, then this simply becomes

`bool nonEmpty = left < right ? runtParser(leftBlubbb, rightBlubb, false) : runParser(right, left, true);`